### PR TITLE
create an automl option to remove unnecessary dependency for autogen and tune

### LINKED ISSUE #725: use packaging.version for version comparison
as suggested by @Yard1 in #721

### LINKED ISSUE #898: Can `xgboost`, `lightgbm`, and `pandas` be made into optional dependencies?
### Discussed in https://github.com/microsoft/FLAML/discussions/897

<div type='discussions-op-text'>

<sup>Originally posted by **topher-lo** January 26, 2023</sup>
Hi `flaml` community and maintainers! First of all, fantastic library and incredible algorithms. My team is using `flaml.tune` to run our custom hyperparameter tuning workflows. It's economical, easy-to-use, fast, and powerful: all 💯 features for production.

There is is one problem causing us headaches though: our Docker images with `flaml` installed are too big. The culprit? `xgboost` and `lightgbm`: we aren't using these models in our ML workflow. Having to install these two dependencies just to use `flaml.tune` is really slowing down our development cycles. Moreover, we are paying for those extra unused MiBs to ECR.

Please correct me if I'm mistaken, but I don't think `lightgbm` and `xgboost` are necessary for `flaml.tune`?

P.S.
Perhaps `pandas` can also be made into an optional dependency? I know many serious data teams that have pure numpy ML workflows in production: `pandas` is just unnecessary bloat. Once again, I don't think `pandas` is required for `flaml.tune`?</div>

### PR DESCRIPTION
**Pull request by sonichi**
## Why are these changes needed?

close #898; close #725 
The new autogen package added one more reason to making an [automl] option and reduce the minimal dependency.

## Checks

- [x] I've used [pre-commit](https://microsoft.github.io/FLAML/docs/Contribute#pre-commit) to lint the changes in this PR, or I've made sure [lint with flake8](https://github.com/microsoft/FLAML/blob/816a82a1155b4de4705b21a615ccdff67c6da379/.github/workflows/python-package.yml#L54-L59) output is two 0s.
- [x] I've included any doc changes needed for https://microsoft.github.io/FLAML/. See https://microsoft.github.io/FLAML/docs/Contribute#documentation to build and test documentation locally.
- [ ] I've added tests (if relevant) corresponding to the changes introduced in this PR.
- [ ] I've made sure all auto checks have passed.

### CODE REVIEW

File: flaml/automl/task/task.py [hunk 1]
```diff
@@ -1,17 +1,11 @@
 from abc import ABC, abstractmethod
 from typing import TYPE_CHECKING, List, Optional, Tuple, Union
-
 import numpy as np
-import pandas as pd
+from flaml.automl.data import DataFrame, Series
```
Comment (thinkall) [inline flaml/automl/task/task.py:4]: Maybe `from flaml.automl.data import psDataFrame, psSeries, DataFrame, Series` and add `psDataFrame, psSeries` to below code.

File: flaml/automl/task/task.py [hunk 2]
```diff
@@ -80,8 +74,8 @@ class Task(ABC):
     def __init__(
         self,
         task_name: str,
-        X_train: Optional[Union[np.ndarray, pd.DataFrame]] = None,
-        y_train: Optional[Union[np.ndarray, pd.DataFrame, pd.Series]] = None,
+        X_train: Optional[Union[np.ndarray, DataFrame]] = None,
```
Comment (thinkall) [inline flaml/automl/task/task.py:77]:
```
        X_train: Optional[Union[np.ndarray, DataFrame, psDataFrame]] = None,
        y_train: Optional[Union[np.ndarray, DataFrame, psDataFrame, Series, psSeries]] = None,
```
Would it be better to update here and in the following type hints?

File: flaml/automl/task/task.py [hunk 3]
```diff
@@ -104,8 +98,8 @@ def evaluate_model_CV(
         self,
         config: dict,
         estimator: "flaml.automl.ml.BaseEstimator",
-        X_train_all: Union[np.ndarray, pd.DataFrame],
-        y_train_all: Union[np.ndarray, pd.DataFrame, pd.Series],
+        X_train_all: Union[np.ndarray, DataFrame],
```
Comment (thinkall) [inline flaml/automl/task/task.py:101]: Update type hints?

File: flaml/automl/task/task.py [hunk 4]
```diff
@@ -136,12 +130,12 @@ def validate_data(
         self,
         automl: "flaml.automl.automl.AutoML",
         state: "flaml.automl.state.AutoMLState",
-        X_train_all: Union[np.ndarray, pd.DataFrame, None],
-        y_train_all: Union[np.ndarray, pd.DataFrame, pd.Series, None],
-        dataframe: Union[pd.DataFrame, None],
+        X_train_all: Union[np.ndarray, DataFrame, None],
```
Comment (thinkall) [inline flaml/automl/task/task.py:133]: Same, update type hints?

File: flaml/automl/task/task.py [hunk 5]
```diff
@@ -169,8 +163,8 @@ def validate_data(
     def prepare_data(
         self,
         state: "flaml.automl.state.AutoMLState",
-        X_train_all: Union[np.ndarray, pd.DataFrame],
-        y_train_all: Union[np.ndarray, pd.DataFrame, pd.Series, None],
+        X_train_all: Union[np.ndarray, DataFrame],
```
Comment (thinkall) [inline flaml/automl/task/task.py:166]: Update type hints?

File: notebook/automl_classification.ipynb [hunk 6]
```diff
@@ -39,7 +39,7 @@
    "metadata": {},
    "outputs": [],
    "source": [
-    "%pip install flaml[notebook]==1.1.2"
+    "%pip install flaml[notebook]==1.1.2 matplotlib openml==0.10.2"
```
Comment (thinkall) [inline notebook/automl_classification.ipynb:42]:
```suggestion
    "%pip install flaml[automl]==1.2.3 matplotlib openml==0.10.2"
```
And the `from flaml.data import CLASSIFICATION` in section  "4. Customized Learner" should be updated to `from flaml.automl.task.task import CLASSIFICATION`

File: notebook/automl_lightgbm.ipynb [hunk 7]
```diff
@@ -39,7 +40,7 @@
    "metadata": {},
    "outputs": [],
    "source": [
-    "%pip install flaml[notebook]==1.0.10"
+    "%pip install flaml[automl]==1.0.10 matplotlib openml==0.10.2"
```
Comment (thinkall) [inline notebook/automl_lightgbm.ipynb:43]: No `automl` option in earlier versions.
```suggestion
    "%pip install flaml[automl]==1.2.3 matplotlib openml==0.10.2"
```

File: notebook/automl_time_series_forecast.ipynb [hunk 8]
```diff
@@ -156,7 +157,7 @@
     }
    ],
    "source": [
-    "%pip install flaml[notebook,ts_forecast]==1.1.2\n",
+    "%pip install flaml[automl,ts_forecast]==1.1.2 matplotlib openml==0.10.2\n",
```
Comment (thinkall) [inline notebook/automl_time_series_forecast.ipynb:160]:
```suggestion
    "%pip install flaml[automl,ts_forecast]==1.2.3 matplotlib openml==0.10.2\n",
```

File: notebook/automl_xgboost.ipynb [hunk 9]
```diff
@@ -39,7 +40,7 @@
    "metadata": {},
    "outputs": [],
    "source": [
-    "%pip install flaml[notebook]==1.1.2"
+    "%pip install flaml[automl]==1.1.2 matplotlib openml==0.10.2"
```
Comment (thinkall) [inline notebook/automl_xgboost.ipynb:43]:
```suggestion
    "%pip install flaml[automl]==1.2.3 matplotlib openml==0.10.2"
```

File: notebook/automl_classification.ipynb (hunk 6 above)
Comment (thinkall) [reply notebook/automl_classification.ipynb:42]: Also, maybe it's a good time to update and check all the notebooks as mentioned in #1012

We need to test each of the notebooks with a new virtual env to make sure it works as expected.

File: website/docs/Examples/Integrate - Scikit-learn Pipeline.md [hunk 10]
```diff
@@ -1,6 +1,11 @@
 As FLAML's AutoML module can be used a transformer in the Sklearn's pipeline we can get all the benefits of pipeline.
 
-This example requires openml==0.10.2.
+### Prerequisites
+
+Install the [automl] option.
+```bash
+pip install "flaml[automl] openml==0.10.2"
```
Comment (thinkall) [inline website/docs/Examples/Integrate - Scikit-learn Pipeline.md:7]: We do need to update the version of openml as in conda forge, the oldest version is 0.11.0

https://anaconda.org/conda-forge/openml/files

**Review by thinkall — CHANGES_REQUESTED** [review]

File: notebook/automl_classification.ipynb (hunk 6 above)
Comment (sonichi) [reply notebook/automl_classification.ipynb:42]: > Also, maybe it's a good time to update and check all the notebooks as mentioned in #1012
> 
> We need to test each of the notebooks with a new virtual env to make sure it works as expected.

That'll be nice. Some notebooks require GPUs though, which will be hard to automate.

**Comment by sonichi** [comment]
Addressed the comments except adding notebook tests. Reasons explained. Hope we can find a solution in future.

**Review by thinkall — APPROVED** [review]
Left a comment for synapse option. Approved for not blocking the merge.

**Review by qingyun-wu — APPROVED** [review]

**Pull request merged**